[IMPROVE] Add departmentName as query param and route to that department in SMS endpoint

bhardwajaditya opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->
Closes https://github.com/WideChat/Rocket.Chat/issues/914
<!-- Your Pull Request name should start with one of the following tags
  [NEW] For new features
  [IMPROVE] For an improvement (performance or little improvements) in existing features
  [FIX] For bug fixes that affect the end-user
  [BREAK] For pull requests including breaking changes
  Chore: For small tasks
  Doc: For documentation
-->

<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)
<!-- CHANGELOG -->
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description will appear in the release notes if we accept the contribution.
-->

- Added department name as query param In the SMS endpoint. Department name should be URL safe.
- Updated SMS endpoint description in admin settings

<img width="976" alt="image" src="https://user-images.githubusercontent.com/23612313/140186113-a3c3b670-c14a-4b92-933c-91d8bc577e57.png">


<!-- END CHANGELOG -->

## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->

## Steps to test or reproduce
Modify the post endpoint URL in twilio WhatsApp configuration as https://<SERVER_URL>/api/v1/livechat/sms-incoming/twilio?departmentName=< departmentName> to test
<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

bhardwajaditya committed:
Add department ID as query param and route to that department in SMS endpoint

ear-dev commented:
@bhardwajaditya just to clarify please - when the new parameter 'departmentID' is present, it will override the default setting in the admin UI?  But if the attribute is not present, it will default to the configured department?  And what form is the departmentID take?  Is it a department name, or the actual Rocketchat ID found somewhere in the database?

bhardwajaditya commented:
@ear-dev yes it's would override the setting value, and will fallback to the default value if it's not present. This is the department name, the same thing what we set in the setting

ear-dev commented:
@bhardwajaditya if my departmentID is "Viasat Customer Support".....  how does that need to be formatted to be used as an attribute in the URL?  Does it have to be something like "Viasat%20Customer%20Support"?  Or do we use quotation marks? or?

bhardwajaditya commented:
@ear-dev yes it needs to be in URL safe form, no quotations. just Viasat%20Customer%20Support

ear-dev commented:
@bhardwajaditya can you catch up with develop_pwa in this PR please?  I think we should be testing all the reusable CX stuff against the latest 4.x code..  thanks.

bhardwajaditya committed:
Merge branch 'develop_pwa' of https://github.com/WideChat/Rocket.Chat into sms_endpoint_department_change

bhardwajaditya commented:
@ear-dev done

ear-dev commented:
@bhardwajaditya So far in my testing this feature is working great, but it is not overriding the default. There seems to be two defaults in play..... "Admin/SMS/Omnichannel Department (Default)" and "Admin/Omnichannel/Routing/Assign new conversations to department".

I think it seems to me that if either one of those is set, then the visitor is routed to that default department. Can you verify if you are seeing the same thing please?

I think we would like to have both default values set, but if the departmentID parameter is in the SMS endpoint, it should prevail. Thanks.

ear-dev reviewed (approved):
Works great!  thanks.

ear-dev commented:
@bhardwajaditya can you add something in the description field under "SMS/Omnichannel Department (Default)" that indicates how you can use the new parameter to route to a specific department please?

Also, can you please create an Issue in the upstream RC repo that describes this feature, and then make an upstream PR for it?  Renato would like to include this, and I'd like to do him the favor if you don't mind making the upstream PR.  Thanks.

Shailesh351 reviewed (approved):
Looks good to me

Shailesh351 commented on app/livechat/imports/server/rest/sms.js:
```diff
@@ -70,8 +70,9 @@ API.v1.addRoute('livechat/sms-incoming/:service', {
 	post() {
 		const SMSService = SMS.getService(this.urlParams.service);
 		const sms = SMSService.parse(this.bodyParams);
+		const { departmentID } = this.queryParams;
```
@bhardwajaditya Is this query param departmentId (i.e. `pDSPNogBNMGv5iAtX `) or departmentName(i.e. `Viasat Customer Support`) ?

If it's `Viasat Customer Support` we should rename it to `departmentName` to avoid confusion

bhardwajaditya replied:
Changed this.

ear-dev commented:
@bhardwajaditya Just for tracking.......  work still to be done on this...... 

- link rooms to visitorToken+departmentID
- document with a note in the adminUI/sms how to use the departmentID attribute in the sms endpoint call to route conversations.

bhardwajaditya committed:
Link SMS sessions to visitor token and department ID

bhardwajaditya committed:
Add SMS endpoint description in admin settings

bhardwajaditya commented:
@ear-dev added the following changes to this PR.

- [X] departmentID changes to departmentName as suggested by Shailesh. Updated the usage in description as well
- [X] Rooms will be searched by visitor token and department ID instead of just visitor token as was the case earlier
- [X] Added description in admin settings for the API usage.

ear-dev commented:
This is looking good @bhardwajaditya I will merge.
